subctl/issue-196-disbale-features-in-build.md

```diff
@@ -0,0 +1,88 @@
+# Disable certain features while building `subctl`
+
+Related Issue:
+[Support disabling certain features in builds](https://github.com/submariner-io/enhancements/issues/196)
+
+## Summary
+
+Some builds of `subctl` (e.g. for OCM) don’t need certain sub-commands such as `deploy-broker`, `join` etc. It would be useful to be able
+to disable these features so that users don’t even see them in `--help` output.
+
+## Proposal
+
+Go provides [build constraints](https://pkg.go.dev/cmd/go#hdr-Build_constraints) to identify files that should be included in the package.
+Build constraints are given by a line comment that begins
+
+    //go:build
+
+Constraints may appear in any kind of source file (not just Go), but they must appear near the top of the file, preceded only by blank
+lines and other line comments. These rules mean that in Go files a build constraint must appear before the package clause.
+
+To distinguish build constraints from package documentation, a build constraint should be followed by a blank line.
+
+A build constraint comment is evaluated as an expression containing build tags combined by ||, &&, and ! operators and parentheses.
+Operators have the same meaning as in Go.
+
+All the sub-commands that are related to deployment would be ignored from specific builds by adding build constraints.
+
+## Design
+
+Below `subctl` sub-commands would be having a build constraint so that they could be added in the package based on the user preferences.
+
+    deploy-broker
+    join
+    recover-broker-info
+    All cloud prepare
+
+The build directive would be `!troubleshoot`. An alternative directive could be `!non-deploy`
```

Comment:
`troubleshoot` is a good idea, but let’s describe it from the tag rather than the build directive (starting with a negated term makes the description more difficult to get right). The executable we want to end up with when using tags contains only troubleshooting and service exporting features. So we need to find something which describes that when used in

    go build -tags foo

(I don’t have a good replacement right now.)

Reply:
The idea is to not change the upstream build but provide mechanism to ignore deploy related commands. The upstream build will be the vanilla build with all the commands and the build mechanism (go build command) to be run without any build tags.

This proposal just adds NEGATED `nondeploy` constraints to build command so that the downstream build (taken as an example to explain the use case) will be built using `--tags nondeploy` so that the deploy related commands are excluded.

So the upstream build mechanism is not changed. Build changes happen only for downstream builds where we dont want any commands. These `Makefile` changes will be made only downstream.

Reply:
One of the goals I had in mind was something we’d talked about a while ago, supporting `go install` — so that means ensuring that the default constraints produce a complete `subctl`. I know the current naming isn’t great...